### ISSUE TITLE
Dynamic data model: Default values

### ISSUE THREAD

**Issue by HeyJoel** [cofoundry-cms/cofoundry#181]
It could be useful to be able to display a default value for a data model field. My situation was needing to set the numerical value of a *NumColumns* field to 3.

This could be done either with a data annotation, or we could somehow get the default values of the field from a model instance which would allow you to use the class constructor (as you normally would for default values in .NET).

One tricky area might be if you wanted to set a date field to the current date. At the moment I've always considered data model schemas to be static (not that they are cached yet), but the current date could change.

**Comment by mcockrellsana** [comment]
Is this currently possible?  I'm trying to figure out how to set default values on custom entities but neither the DefaultValue attribute or setting the value in the constructor seem to work.  Any ideas?

**Comment by HeyJoel** [comment]
Hi there, the issue is open and the feature has not been developed yet.

I wasn't aware there was a `DefaultValueAttribute` already in 'System.ComponentModel', so we'll have to investigate that attribute when we look at developing this feature.

**Comment by mcockrellsana** [comment]
Hey Joel, thanks for the response!  Looking into the CMS code a bit, I can sort of see the area(s) this would need to be changed/updated to pull this attribute in and associate it.  Is there anyway off the top of your head you know the areas that would be affected by this, mainly the methods in C# that gather this information and/or the angular sections which this would apply?  I would love to be able to add this feature ourselves since it's a big requirement on my end, sounds like it might take awhile to get this baked into the cms itself.  Or on another note, is there a way for me to override specific things within our own project to accomplish this?

I appreciate any help on the matter!

**Comment by mcockrellsana** [comment]
I actually asked the question before looking too far into it myself, mistake, I apologize.  So the `AdditionalAttributes` property on the metadata does not see DefaultValue, however, this **is** present on the `propertyMetaData.Attributes` and can probably be pulled from there, updating the `DynamicDataModelSchemaProperty` to store this data.  My bigger question now would be, modifying this, without searching a ton of files for usage, do you know where this would have to be changed/used to accommodate, I'm guessing that this will mainly only affect the angular side after this point, but is there a ton of places this information is used/passed to?

**Comment by HeyJoel** [comment]
On the C# side it's pretty easy to override any of our interfaces (e.g. `IDynamicDataModelSchemaMapper` using the [DI System](https://www.cofoundry.org/docs/framework/dependency-injection). However I don't think you'd want to override that, instead you can use `IMetadataAttribute` (or  `IModelMetadataDecorator` for existing annotations) to pass additional meta data through to the admin UI.

The best thing to do here is look at the existing [MetaDataAttributes](https://github.com/cofoundry-cms/cofoundry/tree/master/src/Cofoundry.Domain/Domain/ModelMetadata/Attributes) and [ModelMetaDataDecorators](https://github.com/cofoundry-cms/cofoundry/tree/master/src/Cofoundry.Domain/Domain/ModelMetadata/ModelMetadataDecorators).

As you notes in your follow-up message, it might be that the existing `DefaultValueAttribute` is pulling through the data you need already.

### JS

On the admin panel (AngularJS) side things are a little tricker to override. In .NET Framework originally had a great system whereby dropping a js file in your local project in the exact location of the source file in the original project would cause it to dynamically compile with your local version. Since .NET Core removed the bundling and minification feature we has to use grunt to pre-build the scripts.

Your best bet now is probably to either to 

1. copy down the entire `Cofoundry.Web.Admin` project and reference that and change the files you need. You'd need to reference the [Cofoundry.Web](https://www.nuget.org/packages/Cofoundry.Web/) package rather than Cofoundry.Web.Admin.
2. copy down just the js files in the [`Admin` and `Build` folders](https://github.com/cofoundry-cms/cofoundry/tree/master/src/Cofoundry.Web.Admin/Admin) from the `Cofoundry.Web.Admin` project into your local solution and rely on the local-override behaviour of the updated files. You'll also need to remove any .cs files from these folders as they will conflict with your references admin package.

Either way you'd need to rebuild the js project using the grunt script in the `Build` directory. Sorry there's not much information on how to do this, we haven't had time to document any of the admin panel JS stuff yet.

As for the code to look into, try these:

- [FormDynamicFieldSet](https://github.com/cofoundry-cms/cofoundry/tree/master/src/Cofoundry.Web.Admin/Admin/Modules/Shared/Js/UIComponents/FormDynamicFieldSet): Used to render dynamic forms from the .NET metadata information returned from the API
- [FormFields](https://github.com/cofoundry-cms/cofoundry/tree/master/src/Cofoundry.Web.Admin/Admin/Modules/Shared/Js/UIComponents/FormFields): A folder containing most of the standard form fields displayed by the field set

**Comment by mcockrellsana** [comment]
Wow, thank you so much for the comprehensive response.  I will get to work on this and update you once I've completed this.  Once I do, i'll also fork the current master (unless you have different process in place) and make changes and add a PR for you so you don't have to worry about this in the future!

**Comment by mcockrellsana** [comment]
I will note that I have the master branch of coFoundry as a git submodule in my main project, so we are directly referencing your code and projects within our own.  Would this allow me to use the override behavior of the JS files since i'm not using nuget packages?

**Comment by HeyJoel** [comment]
If you're referencing the code directly rather than the NuGets, then you can just update the admin code directly (re-running the js build of course), rather than relying on the override behaviour, although it will still work. Basically the static resource provider in Cofoundry will check for local files first before resolving to embedded resources.

Any files from other projects get compiled as an embedded resource, so by default when developing on js files in the admin project you'll need to re-compile to see the change in your local project. There's a few [config settings](https://www.cofoundry.org/docs/references/common-config-settings#debugsettings) that allow you to point to those embedded files directly and avoid the re-compilation step, which is really useful when making changing to the admin code.

With regards to a PR, that would be most welcome, although I still need to set up a CLA and get you to sign it before we can accept the PR. This is on my list of things to do and will sort it out if you do submit the PR. You can fork master, but do use a feature branch.

**Comment by HeyJoel** [comment]
This is fixed and will be released in 0.9. You can assign default property values in your model definition and they will be used int he admin UI e.g:

```csharp
public class HeaderBlock : INestedDataModel
{
    [Required]
    public string Text { get; set; } = "Default Title";
}
```

This works with any of the data model types, i.e. custom entities, page blocks and nested data models.
